derive/src/parse.rs

```diff
@@ -462,59 +462,121 @@ fn next_enum(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Enum
                     },
                     "," => {
-                        ret.insert(typename.clone(), generic_bound.clone());
+                        ret.push((typename.clone(), std::mem::take(&mut generic_bounds)));
                         typename.clear();
-                        generic_bound.clear();
                         in_type = true
                     },
                     c => {
                         if in_type {
                             typename+=c;
                         } else {
-                            generic_bound += c;
+                            generic_bounds.push(c.to_owned());
                         }
                     }
                 }
             }
         },
         _ => ()
     }
+
+    // "where" generics + bounds
+    if let Some(content) = source.peek() {
+        if content.to_string() != "where"  {
+            return ret;
+        } else {
+            source.next();
+        }
+
+        let mut typename = "".to_string(); 
+        let mut bound = "".to_string();
+        let mut generic_bounds = Vec::new();
+        let mut in_type = true;
+        while let Some(tok) = source.peek() {
+            match tok.to_string().as_str() {
+                end if end.starts_with("{") => {
+                    if typename.is_empty() { break;}
+                    if let Some(entry) = ret.iter_mut().find(|x| typename == x.0) {
+                        if !bound.is_empty() {
+                            generic_bounds.push(std::mem::take(&mut bound));
+                        }
+                        entry.1.extend((std::mem::take(&mut generic_bounds)).into_iter());
+                    } else {
+                        panic!("Generics in where bounds must be previously declared in <angle brackets>")
+                    }
+                    break
+                },
+                colon @ ":" => {
+                    if in_type {
+                        in_type = false;
+                    } else {
+                        bound+=colon;
+                    }
+                    
+                },
+                "," => {
+                    if !bound.is_empty() {
+                        generic_bounds.push(std::mem::take(&mut bound));
+                    }
+                    if let Some(entry) = ret.iter_mut().find(|x| typename == x.0) {
+                        entry.1.extend((std::mem::take(&mut generic_bounds)).into_iter());
+                    } else {
+                        panic!("Generics in where bounds must be previously declared in <angle brackets>")
+                    }
+                    typename.clear();
+                    in_type = true
+                },
+                "+" => {
+                    if !bound.is_empty() {
+                        generic_bounds.push(std::mem::take(&mut bound));
+                    }
+                },
+                c => {
+                    if in_type {
+                        typename+=c;
```

Comment:
you probably want to run `rustfmt` on this as well :) 

it will make this line `typename += c;`

Reply:
ran this